lib/vector-core/buffers/benches/common.rs

```diff
@@ -0,0 +1,177 @@
+use buffers::bytes::{DecodeBytes, EncodeBytes};
+use buffers::{self, Variant};
+use bytes::{Buf, BufMut};
+use futures::task::{noop_waker, Context, Poll};
+use futures::{Sink, Stream};
+use std::fmt;
+use std::pin::Pin;
+
+#[derive(Clone, Copy)]
+pub struct Message<const N: usize> {
+    id: u64,
+    _padding: [u64; N],
+}
+
+impl<const N: usize> Message<N> {
+    fn new(id: u64) -> Self {
+        Message {
+            id,
+            _padding: [0; N],
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum EncodeError {}
+
+#[derive(Debug)]
+pub enum DecodeError {}
+
+impl fmt::Display for DecodeError {
+    fn fmt(&self, _f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        unreachable!()
+    }
+}
+
+impl<const N: usize> EncodeBytes<Message<N>> for Message<N> {
+    type Error = EncodeError;
+
+    fn encode<B>(self, buffer: &mut B) -> Result<(), Self::Error>
+    where
+        B: BufMut,
+        Self: Sized,
+    {
+        buffer.put_u64(self.id);
+        for _ in 0..N {
+            // this covers self._padding
+            buffer.put_u64(0);
+        }
+        Ok(())
+    }
+}
+
+impl<const N: usize> DecodeBytes<Message<N>> for Message<N> {
+    type Error = DecodeError;
+
+    fn decode<B>(mut buffer: B) -> Result<Self, Self::Error>
+    where
+        B: Buf,
+        Self: Sized,
+    {
+        let id = buffer.get_u64();
+        for _ in 0..N {
+            // this covers self._padding
+            let _ = buffer.get_u64();
+        }
+        Ok(Message::new(id))
+    }
+}
+
+#[allow(clippy::type_complexity)]
+pub fn setup<const N: usize>(
+    max_events: usize,
+    variant: Variant,
+) -> (
+    Box<dyn Sink<Message<N>, Error = ()> + Unpin + Send>,
+    Box<dyn Stream<Item = Message<N>> + Unpin + Send>,
```

Comment:
TIOLI: The box overhead is probably minimal in reality, but… could this not be rejiggered in the `impl Trait` fashion to avoid boxing?

Reply:
Well spotted. I think it can / should but it's a change that quickly spirals out to the rest of the project. Well worth creating a follow-up ticket.